paramiko/sftp_file.py

```diff
@@ -522,18 +530,26 @@ def _get_size(self):
         except:
             return 0
 
-    def _start_prefetch(self, chunks):
+    def _start_prefetch(self, chunks, max_outstanding_requests = None):
         self._prefetching = True
         self._prefetch_done = False
 
-        t = threading.Thread(target=self._prefetch_thread, args=(chunks,))
+        t = threading.Thread(target=self._prefetch_thread, args=(chunks,max_outstanding_requests))
         t.daemon = True
         t.start()
 
-    def _prefetch_thread(self, chunks):
+    def _prefetch_thread(self, chunks, max_outstanding_requests):
         # do these read requests in a temporary thread because there may be
         # a lot of them, so it may block.
         for offset, length in chunks:
+            # Limit the number of outstanding requests in a busy-loop
+            if max_outstanding_requests is not None:
+                while True:
+                    with self._prefetch_lock:
+                        if len(self._prefetch_extents) < max_outstanding_requests:
```

Comment:
Checking my understanding of the implementation here -- does each call to `self.sftp._async_request()` cause `self._prefetch_extents` to increment by one, and then once each chunk is received, `._prefetch_extents` is decremented?

And, thus, this new code causes this queuing within `self._prefetch_thread()` to idle whenever we hit our max-requests cap?

Reply:
The new async request number returned by `_async_request` is added to `_prefetch_extents` by the code below (hidden by the diff view on github by default:

```
            with self._prefetch_lock:
                self._prefetch_extents[num] = (offset, length)
```

The request number is then removed again in `_async_response`.

So yes, the background thread just idles here if the limit has been reached.